### PR TITLE
Adding old objectives

### PR DESCRIPTION
**Pull request by unalmis**
until the nan in forward for GammaC is resolved (1489) and nuffts are used (#1294)

### CODE REVIEW

File: desc/objectives/_neoclassical.py [hunk 1]
```diff
@@ -283,3 +283,120 @@ def compute(self, params, constants=None):
             **self._hyperparam,
         )
         return constants["transforms"]["grid"].compress(data["effective ripple"])
+
+
+class OldEffectiveRipple(EffectiveRipple):  # noqa: D101
```
Comment (dpanici) [inline desc/objectives/_neoclassical.py:288]: maybe keep it a single class but add a `interpolation_method` flag to be `1D` or `2D`

Comment (unalmis) [reply desc/objectives/_neoclassical.py]: that would be clean but we did not know if combining the objectives would have disadvantage for jit compilation

File: desc/objectives/_fast_ion.py [hunk 2]
```diff
@@ -295,3 +303,105 @@ def compute(self, params, constants=None):
             **self._hyperparam,
         )
         return constants["transforms"]["grid"].compress(data[self._key])
+
+
+class GammaC_Spline(GammaC):  # noqa: D101
```
Comment (dpanici) [inline desc/objectives/_fast_ion.py:308]: use a flag to choose method

**Comment by f0uriest** [comment]
This would also be a good place to add benchmarks for the bounce integral objectives (compute and jacobian, ideally using both 1d and 2d versions)

File: desc/compute/_old.py
Comment (f0uriest) [inline desc/compute/_old.py:1]: maybe something like `_neoclassical_1d.py` rather than just `old` or `deprecated`

**Comment by unalmis** [comment]
> This would also be a good place to add benchmarks for the bounce integral objectives (compute and jacobian, ideally using both 1d and 2d versions)

That is additional work for a separate pull request. You'd need to scan through different parameter settings (num alpha, num transit, num quad, eq.M, eq.N). Also compare forward, reverse and the current jax issues prevent that. Then re-run and reinterpret the results with with each of the remaining major updates, #1154, #1294, #1303, #1206, #1034, and the two open PRs for faster transforms and partial summation/vandermonde stuff.

**Comment by f0uriest** [comment]
> > This would also be a good place to add benchmarks for the bounce integral objectives (compute and jacobian, ideally using both 1d and 2d versions)
> 
> That is additional work for a separate pull request. You'd need to scan through different parameter settings (num alpha, num transit, num quad, eq.M, eq.N). Also compare forward, reverse and the current jax issues prevent that. Then re-run and reinterpret the results with with each of the remaining major updates, #1154, #1294, #1303, #1206, #1034, and the two open PRs for faster transforms and partial summation/vandermonde stuff.

I don't think we need to scan over every parameter, the default for each objective is fine. But it would be useful to have some sort of benchmark before doing those other improvements so we know what helps and what doesn't

**Comment by unalmis** [comment]
> I don't think we need to scan over every parameter, the default for each objective is fine. But it would be useful to have some sort of benchmark before doing those other improvements so we know what helps and what doesn't

#1575 can be merged here upon review

File: tests/benchmarks/benchmark_cpu_small.py [hunk 3]
```diff
@@ -458,3 +459,58 @@ def run(obj, con):
...
+        ]
+    )
+    objective.build(eq)
+    objective.compile()
+    x = objective.x(eq)
+
+    def run(x, objective):
+        objective.compute_scaled_error(x, objective.constants).block_until_ready()
+
+    benchmark.pedantic(run, args=(x, objective), rounds=3, iterations=1)
```
Comment (YigitElma) [inline tests/benchmarks/benchmark_cpu_small.py:488]: If this doesn't take really long time, I would suggest to increase the number of rounds to at least 10 ish. Github runners seems to be noisy.

File: tests/benchmarks/benchmark_cpu_small.py [hunk 4]
```diff
@@ -458,3 +459,58 @@ def run(obj, con):
...
```
Comment (YigitElma) [inline tests/benchmarks/benchmark_cpu_small.py:516]: Same as above

File: tests/benchmarks/benchmark_cpu_small.py [hunk 5]
```diff
@@ -458,3 +459,58 @@ def run(obj, con):
         rounds=10,
         iterations=1,
     )
+
+
+@pytest.mark.slow
+@pytest.mark.benchmark
```
Comment (YigitElma) [inline tests/benchmarks/benchmark_cpu_small.py:465]: We don't run the gpu benchmarks on CI but still we try to keep cpu and gpu files the same. Can you add these to that file too? When you commit you need to put `--no-verify` to prevent pre-commit from complaining `set_device("gpu")` line.

File: tests/benchmarks/benchmark_cpu_small.py [hunk 6]
```diff
@@ -458,3 +459,60 @@ def run(obj, con):
...
+            EffectiveRipple(
+                eq,
+                num_transit=num_transit,
+                num_well=10 * num_transit,
+                num_quad=16,
+            )
+        ]
+    )
+    objective.build(eq)
+    objective.compile(mode="bfgs")
```
Comment (YigitElma) [inline tests/benchmarks/benchmark_cpu_small.py:483]: `bfgs` will compile `objective.compute_scalar` and `objective.grad`, so, in this case compiling doesn't help. Personally, I favor calling `objective.compute_scaled_error(x, objective.constants).block_until_ready()` here, instead of `objective.compile`. One, it prevents this kind of confusion, and two, it doesn't spend time compiling the derivatives.

Comment (unalmis) [reply tests/benchmarks/benchmark_cpu_small.py:483]: I thought we were timing this?

Comment (YigitElma) [reply tests/benchmarks/benchmark_cpu_small.py:483]: It depends on what you want to profile. For other benchmarks with the same structure, the part we time is individual calls to `objective.compute_scaled_error(x, objective.constants).block_until_ready()`, so only the function `run`. Since in real applications, the compute function is compiled once, we profile the following runtimes. If you want to also benchmark how long it takes to compile the function, add `jax,clear_caches()` inside `run` before the `compute` call.

Comment (unalmis) [reply tests/benchmarks/benchmark_cpu_small.py:483]: i'll add objective.grad to run

Comment (YigitElma) [reply tests/benchmarks/benchmark_cpu_small.py:483]: The current version will call `run` 10 times, and the first one will take very long since it also needs to compile, but the other 9 calls will be only the evaluation of compiled `compute`. If you look at the details, you can see this,
image
For printing benchmarks, we use the median value, hence this won't change things much, but still having all rounds do the same thing is good practice.

File: tests/benchmarks/benchmark_cpu_small.py [hunk 7]
```diff
@@ -458,3 +459,60 @@ def run(obj, con):
...
+
+
+@pytest.mark.slow
+@pytest.mark.benchmark
+def test_ripple_objective_1D(benchmark):
+    """Benchmark computing objective for effective ripple."""
+    eq = desc.examples.get("W7-X")
+    with pytest.warns(UserWarning, match="Reducing radial"):
+        eq.change_resolution(L=eq.L // 2, M=eq.M // 2, N=eq.N // 2)
+    num_transit = 5
```
Comment (YigitElma) [inline tests/benchmarks/benchmark_cpu_small.py:499]: Also, if you think this is unrealistically low res, you can increase it. The benchmarks say that `compute` takes 80ms, we might not see improvements in this resolution.

File: tests/benchmarks/benchmark_cpu_small.py (hunk 6 above)
Comment (YigitElma) [reply tests/benchmarks/benchmark_cpu_small.py:483]: Instead of `objective.grad` maybe use similar thing like `test_proximal_freeb_compute` and `test_proximal_freeb_jac`. Those are the actual derivatives taken during the optimization (not including the proximal.eq_update` part.

Comment (unalmis) [reply tests/benchmarks/benchmark_cpu_small.py]: [4be6703](https://github.com/PlasmaControl/DESC/pull/1514/commits/4be67032f64c39af4cf62cdd06c8656fee62ae57)?

Comment (unalmis) [reply tests/benchmarks/benchmark_cpu_small.py]: Why don't we just add benchmarking to the computations in test_neoclassical and test_fast_ion?

**Comment by unalmis** [comment]
The benchmarks on this pull request run fine. However the benchmarks for master are actually running the benchmarks from commit https://github.com/PlasmaControl/DESC/pull/1514/commits/e6d806b167d4d657d84c1fb0a0d2052d33968104 on this branch, which is what fails the CI. 
@YigitElma Do you know why master benchmarks attempt to run a test from a commit that is not even in master?

File: tests/benchmarks/benchmark_cpu_small.py (hunk 6 above)
Comment (YigitElma) [reply tests/benchmarks/benchmark_cpu_small.py]: I was thinking like,
```python
@pytest.mark.slow
@pytest.mark.benchmark
def test_proximal_effective_ripple_jac(benchmark):
    eq = desc.examples.get("ESTELL")
    with pytest.warns(UserWarning, match="Reducing radial"):
        eq.change_resolution(8, 8, 8, 16, 16, 16)
    num_transit = 20
    objective = ObjectiveFunction(
        [
            EffectiveRipple(
                eq,
                num_transit=num_transit,
                num_well=10 * num_transit,
                num_quad=16,
            )
        ]
    )
    constraint = ObjectiveFunction(ForceBalance(eq))
    prox = ProximalProjection(objective, constraint, eq)
    prox.build()
    x = prox.x(eq)
    prox.jac_scaled_error(x, prox.constants).block_until_ready()

    def run(x, prox):
        prox.jac_scaled_error(x, prox.constants).block_until_ready()

    benchmark.pedantic(run, args=(x, prox), rounds=10, iterations=1)


@pytest.mark.slow
@pytest.mark.benchmark
def test_proximal_effective_ripple_compute(benchmark):
    eq = desc.examples.get("ESTELL")
    with pytest.warns(UserWarning, match="Reducing radial"):
        eq.change_resolution(8, 8, 8, 16, 16, 16)
    num_transit = 20
    objective = ObjectiveFunction(
        [
            EffectiveRipple(
                eq,
                num_transit=num_transit,
                num_well=10 * num_transit,
                num_quad=16,
            )
        ]
    )
    constraint = ObjectiveFunction(ForceBalance(eq))
    prox = ProximalProjection(objective, constraint, eq)
    prox.build()
    x = prox.x(eq)
    prox.compute_scaled_error(x, prox.constants).block_until_ready()

    def run(x, prox):
        prox.compute_scaled_error(x, prox.constants).block_until_ready()

    benchmark.pedantic(run, args=(x, prox), rounds=10, iterations=1)


@pytest.mark.slow
@pytest.mark.benchmark
def test_proximal_effective_ripple_compute_compilation(benchmark):
    eq = desc.examples.get("ESTELL")
    with pytest.warns(UserWarning, match="Reducing radial"):
        eq.change_resolution(8, 8, 8, 16, 16, 16)
    num_transit = 20
    objective = ObjectiveFunction(
        [
            EffectiveRipple(
                eq,
                num_transit=num_transit,
                num_well=10 * num_transit,
                num_quad=16,
            )
        ]
    )
    constraint = ObjectiveFunction(ForceBalance(eq))
    prox = ProximalProjection(objective, constraint, eq)
    prox.build()
    x = prox.x(eq)

    def run(x, prox):
        jax.clear_caches()
        prox.compute_scaled_error(x, prox.constants).block_until_ready()

    benchmark.pedantic(run, args=(x, prox), rounds=5, iterations=1)
```

Feel free to change resolutions and stuff. But the main point, if you don't create a `ProximalProjection`, the derivative is computed for each equilibrium parameter R_lmn, Rb_lmn etc. In optimization, we rarely take the full derivative, `ProximalProjection` reduces the derivatives to be only in boundary parameters Rb_lmn. For compute benchmarks, proximal shouldn't change anything, but I just copy/pasted jacobian cases. You can write your own version for that.

Comment (YigitElma) [reply tests/benchmarks/benchmark_cpu_small.py]: > Why don't we just add benchmarking to the computations in test_neoclassical and test_fast_ion?

For compute, you can use one of them I guess.

**Comment by YigitElma** [comment]
> The benchmarks on this pull request run fine. However the benchmarks for master are actually running the benchmarks from commit [4200972](https://github.com/PlasmaControl/DESC/commit/42009726a41659c73d71bd6144401decc31b5acb) on this branch, which is what fails the CI. @YigitElma Do you know why master benchmarks attempt to run a test from a commit that is not even in master?

Yeah, this is expected. Our benchmarks basically run the same `benchmarks_cpu_small.py` file for both master and PR branches. So, we checkout to master inside benchmarks but we keep the `benchmarks_cpu_small.py` the same as the feature branch.

**Comment by unalmis** [comment]
> The benchmarks on this pull request run fine. However the benchmarks for master are actually running the benchmarks from commit [4200972](https://github.com/PlasmaControl/DESC/commit/42009726a41659c73d71bd6144401decc31b5acb) on this branch, which is what fails the CI. @YigitElma Do you know why master benchmarks attempt to run a test from a commit that is not even in master?

I turned on more verbose debugging. What is exactly happening is that master is attempting to run a benchmark added on this branch, however it fails because running that benchmark requires the old objectives added on this branch.

**Comment by YigitElma** [comment]
If you want benchmark some code that doesn't exist in master, you should wait for this one to be merged and then add benchmarks for that unfortunately.

File: tests/benchmarks/benchmark_cpu_small.py (hunk 6 above)
Comment (unalmis) [reply tests/benchmarks/benchmark_cpu_small.py]: The tests fail when I make the suggested change to use proximal instead of just using the raw objective.gradient. Am I missing something?

**Comment by unalmis** [comment]
> If you want to benchmark some code that doesn't exist in master, you should wait for this one to be merged and then add benchmarks for that unfortunately.

In that case, this PR is ready for merge.

**Review by YigitElma — APPROVED** [review]
Ideally, I would like to have benchmarks in a separate PR, but given the high number of existing PRs and since they pass for this PR, I don't think this would cause an issue.

File: desc/compute/_old.py
Comment (unalmis) [reply desc/compute/_old.py:1]: I think in the longer term, once all the todo's are done, these will be removed anyway.

**Review by f0uriest — APPROVED** [review]
